### PR TITLE
Avoid double registration of providers

### PR DESCRIPTION
**Pull request by radcortez**
When `useBuiltinProviders` from RESTEasy is `true`, the `availableProviders` will be automatically discovered by RESTEasy. If `availableProviders` are also included, RESTEasy will warn about double regisration.

### CODE REVIEW

**Comment by gsmet** [comment]
@cdsap The failure reported above look a bit concerning ^

File: extensions/resteasy-classic/resteasy-common/deployment/src/main/java/io/quarkus/resteasy/common/deployment/ResteasyCommonProcessor.java [hunk 1]
```diff
@@ -232,8 +232,13 @@ JaxrsProvidersToRegisterBuildItem setupProviders(BuildProducer<ReflectiveClassBu
...
 
         if (useBuiltinProviders) {
+            contributedProviders.removeAll(availableProviders);
             providersToRegister = new HashSet<>(contributedProviders);
-            providersToRegister.addAll(availableProviders);
+            availableProviders.removeAll(providersToRegister);
+            availableProviders.removeAll(annotatedProviders);
+            for (String availableProvider : availableProviders) {
+                reflectiveClass.produce(ReflectiveClassBuildItem.builder(availableProvider).fields().build());
+            }
```
Comment (gsmet) [inline extensions/resteasy-classic/resteasy-common/deployment/src/main/java/io/quarkus/resteasy/common/deployment/ResteasyCommonProcessor.java:241]: I think this is extremely confusing. If we have to do crazy things like that, we need to categorize the providers properly and not rely on names like `contributedProviders` or `availableProviders`. There's no way to actually figure out what's going on when reading this code without going line by line and try to figure out what they  could contain.

File: extensions/resteasy-classic/resteasy-common/deployment/src/main/java/io/quarkus/resteasy/common/deployment/ResteasyCommonProcessor.java [hunk 2]
```diff
@@ -176,10 +176,10 @@ JaxrsProvidersToRegisterBuildItem setupProviders(BuildProducer<ReflectiveClassBu
             }
         }
         contributedProviders.addAll(annotatedProviders);
+        // this one is added manually in RESTEasy's ResteasyDeploymentImpl
+        contributedProviders.add(ServerFormUrlEncodedProvider.class.getName());
```
Comment (gsmet) [inline extensions/resteasy-classic/resteasy-common/deployment/src/main/java/io/quarkus/resteasy/common/deployment/ResteasyCommonProcessor.java:180]: See my other comment. You're changing the semantic of `contributedProviders` by adding this class in it (it's one that is registered by RESTEasy, not a contributed one) and this is problematic IMO. It makes things extremely confusing.

**Review by gsmet — CHANGES_REQUESTED** [review]
I think this makes things a lot more confusing. I'm not saying the new logic is wrong, just that it's very hard to follow.

We need to make sure the purpose of the sets is understandable and the new code is mixing them all to make it work for what you want to achieve and we completely lost the initial semantic.

I can take a stab at it if you don't feel like it.

**Comment by radcortez** [comment]
> I think this makes things a lot more confusing. I'm not saying the new logic is wrong, just that it's very hard to follow.
> 
> We need to make sure the purpose of the sets is understandable and the new code is mixing them all to make it work for what you want to achieve and we completely lost the initial semantic.
> 
> I can take a stab at it if you don't feel like it.

I agree. I just wanted to minimize the changes. Let me rework this properly.

**Comment by cdsap** [comment]
Hi @gsmet,
after analyzing one of the failing tests, we have for the last [28 days](https://ge.quarkus.io/scans/tests?search.relativeStartTime=P28D&search.timeZoneId=America%2FLos_Angeles&tests.container=io.quarkus.gradle.ImageTasksWithConfigurationCacheTest):
* 1 Flaky execution
* 2 Failing executions

The flaky execution is caused by a `503 Service Unavailable` when building the docker [image](https://ge.quarkus.io/s/fkf4eaitlfl7w/tests/goal/io.quarkus:quarkus-integration-test-gradle-plugin:surefire:test@default-test/details/io.quarkus.gradle.ImageTasksWithConfigurationCacheTest/shouldReuseConfigurationCacheWithProjectIsolationImageBuildIfTheExtensionIsAdded?focused-execution=1&page=eyJvdXRwdXQiOnsiMCI6NH19&top-execution=1#L616) on: 
```
ERROR: failed to solve: registry.access.redhat.com/ubi8/openjdk-21:1.18: failed to resolve source metadata for 
registry.access.redhat.com/ubi8/openjdk-21:1.18: unexpected status from HEAD request to 
https://registry.access.redhat.com/v2/ubi8/openjdk-21/manifests/1.18: 503 Service Unavailable
```

Similarly, the failing executions are related to a `502 Bad Gateway` when requesting `registry.access.redhat.com/ubi8/openjdk-21:1.18`:
```
ERROR: failed to solve: registry.access.redhat.com/ubi8/openjdk-21:1.18: failed to resolve source metadata for registry.access.redhat.com/ubi8/openjdk-21:1.18: unexpected status from HEAD request to 
https://registry.access.redhat.com/v2/ubi8/openjdk-21/manifests/1.18: 502 Bad Gateway
```

What are your thoughts on this? Is the service degradation a temporary issue, or should I avoid building the Docker images during these integration tests?

**Review by gsmet — APPROVED** [review]
I rebased and fixed a small typo, let's merge when CI is green.
